extras/rebalance/directory-rebalance.py

```diff
@@ -0,0 +1,203 @@
+    sys.stdout.flush()
+
+def progress_done():
+    print()
+
+class Rebalancer:
+    def __init__(self, path):
+        self.path = path # Path on which migration script is executed
+        self.migration_duration = 0 #Time spent in migrating data
+        self.skipped_migration_duration = 0 #Time spent in migrating data
+        self.migrated_files = 0 #Number of files migrated successfully
+        self.total_files = 0 #Total number of files scanned
+        self.total_size = 0 #Cumulative size of files scanned so far
+        self.expected_total_size = 0 #This is calculated at the time of populating index
+        self.expected_total_files = 0 #This is calculated at the time of populating index
+        self.migrated_size = 0 #Cumulative size of files migrated
+        self.index = self.get_file_name('index')
+        self.init_logging()
+
+    #Generate a unique name for the given path. format of the path will be
+    #rebalance-<hiphenated-path>-<first-8-hex-chars-of-md5-digest>.suffix
+    #If the length of this name is > 255 then hiphenated-path is truncated to
+    #make space
+    def get_file_name(self, suffix):
+        name_suffix = hashlib.md5(self.path.encode('utf-8')).hexdigest()[:8]+'.'+suffix
+        name_suffix = '-' + name_suffix
+        max_name_len = os.pathconf('.', 'PC_NAME_MAX')
+        name = 'rebalance'+self.path.replace('/', '-')
+        if len(name) > max_name_len - len(name_suffix):
+            name = name[:(max_name_len - len(name_suffix))]
+        name += name_suffix
+        return name
+
+    #Log format is as follows
+    #2020-10-21 18:24:27.838 INFO /mnt/glusterfs/0/aaaaaaaaaa/1 - 1.0 KiB [1024] - 74.6 KiB/s
+    def init_logging(self):
+        logging.basicConfig(filename=self.get_file_name('log'), level=logging.DEBUG,
+                            format='%(asctime)s.%(msecs)03d %(levelname)s %(message)s',
+                            datefmt='%Y-%m-%d %H:%M:%S')
+
+    #Executes the setxattr syscall to trigger migration
+    def migrate_data(self, f):
+        try:
+            os.setxattr(f, "trusted.distribute.migrate-data", b"1",
+                        follow_symlinks=False)
+            return True, None #Indicate that migration happened
+
+        except OSError as e:
+            return False, e
+
+    #Updates the total,migrated,skipped files/size and durations of the migrations
+    def migrate_with_stats(self, f, size):
+        migration_start = time.perf_counter()
+        result, err = self.migrate_data(f)
+        migration_end = time.perf_counter()
+        duration = migration_end - migration_start
+
+        if err is not None:
+            if err.errno == errno.EEXIST:
+                logging.info("{} - Not needed".format(f))
+            elif err.errno == errno.ENOENT:
+                logging.info("{} - file not present anymore".format(f))
+            else:
+                logging.critical("{} - {} - exiting.".format(f, err))
+                self.exit()
```

Comment:
Have you considered using `raise err` to propagate the error instead of forcing a termination ?

Reply:
I can do that too. I will have to break printing of the stats into a separate function, call that and raise the error.

Reply:
If you use the `with` approach, stats can be automatically shown in the `__exit__` method. You don't need to explicitly call it from anywhere, even if there are exceptions.